forge-ai/src/main/java/forge/ai/AiAttackController.java

```diff
@@ -152,6 +147,17 @@ public boolean apply(Card c) {
                     defenders.add(animatedCopy);
                 }
             }
+            // Transform (e.g. Incubator tokens)
+            for (SpellAbility sa : Iterables.filter(c.getSpellAbilities(), SpellAbilityPredicates.isApi(ApiType.SetState))) {
```

Comment:
Looks like this code is pretty much `ComputerUtilCombat..canTransform?`

Reply:
More or less, yes, but sadly, without iterating through spell abilities and grabbing the transform sa (with SetStateAPI) somehow, it'd be impossible to check if the AI can pay the cost for it :/ Do you have any ideas for how to implement that part seamlessly while simplifying the loop (e.g. by changing it into a canTransform check)?

Reply:
I mean, we can check canTransform, and if it returns true, then loop over the SAs and find the SetState one and check the cost, but I'm not sure if that's much better than what it's currently doing :)

Reply:
It looks like it will check `canPayCost` too...?
Otherwise just returns the untransformed

Reply:
The only problem might be we don't add up the costs so the AI might expect a lot of defenders that are actually not possible

Reply:
Yes, true, it's probably also true for other situations similar to this one where the AI weighs every case individually. Not sure how to make it account for all of them together off the top of my head tbh.

Reply:
Hmm, maybe some trickery with remembering the mana sources necessary to pay every processed cost would be possible, but I'm not sure

Reply:
Checked the code for remembering and reserving mana sources, that's only implemented on the AI side (in the AI controller) and hooked up to AI-specific routines... Don't see a way to simulate that for the human opponent in order to determine how much can be paid :( There's probably a roundabout way where we can remember how much mana is already potentially spent (during the loop iteration) and then we bail out of adding defenders like that if we're above the total limit, it's going to be imprecise though.

Reply:
Attempted what I mentioned above, it's imprecise but it sort of works for simple cases like Incubator where only colorless mana is used, I guess. I welcome any better implementation if you can think of one... This is prolly the best I can do for now. I wanted to also hook this up to Animate above, but I'm afraid that'll be a little bit too imprecise since many animate costs depend on particular color shards, and that'll make the thing fail more often than work vs. multicolor decks :/

Reply:
I guess it'll also need to check if the cost actually has a mana part to avoid a NPE?

Reply:
Maybe it can actually also be done for Animate... It'll miscount sometimes, I guess, but won't be as passive when not due? Prolly needs testing though.

Reply:
After some tests, added this to Animate as well and added some NPE prevention. Probably the best I can do for now, I don't have any better ideas, sadly :)